onnxruntime/core/providers/snpe/snpe_lib.cc

```diff
@@ -0,0 +1,401 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/snpe/snpe_lib.h"
+#include <iostream>
+#include <unordered_map>
+#include <memory>
+#include "core/common/common.h"
+#include "core/common/logging/logging.h"
+#include "core/common/safeint.h"
+#include "DlContainer/IDlContainer.hpp"
+#include "DlSystem/ITensorFactory.hpp"
+#include "DlSystem/IUserBufferFactory.hpp"
+#include "SNPE/SNPEBuilder.hpp"
+
+namespace onnxruntime {
+namespace contrib {
+namespace snpe {
+
+size_t CalcSizeFromDims(const zdl::DlSystem::Dimension* dims, size_t rank, size_t elementSize) {
+  if (rank == 0) return 0;
+  SafeInt<size_t> size = elementSize;
+  while (rank--) {
+    size = SafeInt<size_t>(size * (*dims));
```

Comment:
```suggestion
    size = size * (*dims);
```
now that size is already a SafeInt